### PR TITLE
Build OpenAPI Definitions per group instead of per resource

### PR DESCRIPTION
**Pull request by jennybuckley**
**What this PR does / why we need it**:
Addresses inefficiencies in the implementation of #63893

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Part of fixing #66368
Not sure where the 40 seconds comes from in the referenced bug, but this PR reduced startup time of kube apiserver by ~20% on my computer

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go [hunk 1]
```diff
@@ -445,3 +455,80 @@ func NewDefaultAPIGroupInfo(group string, scheme *runtime.Scheme, parameterCodec
...
+					return gvkToSchema, err
+				}
+				gvkToSchema[groupVersion][resource] = openAPISchema
+			}
+		}
+	}
+	return gvkToSchema, nil
+}
+
+func getCanonicalTypeName(model interface{}) string {
```
[the remark this answers is not in the document]
Comment (jennybuckley) [reply staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go]: Yes, this is copied from https://github.com/kubernetes/kube-openapi/blob/master/pkg/builder/openapi.go#L138-L151

**Comment by jennybuckley** [comment]
I just ran some pprof analysis on startup of a basic API server with no extensions, before and after this change.

Before this change, a total of ~16.7 million objects were allocated within the first 50 seconds of startup.
After this change, only ~12.0 million objects were allocated in the same time frame.

**Comment by smarterclayton** [comment]
Hrm, was hoping for a more dramatic reduction.  Where were the bulk of the remaining allocations coming from under a cumulative sort?

**Comment by jennybuckley** [comment]
@smarterclayton About 5.5 million were coming from the openapi endpoint handler, which creates just one full openapi spec and serves it, and this change doesn't touch that code. I think the openapi spec might just take that many objects to create.

**Comment by smarterclayton** [comment]
Ok, we can dig further after this.  Can you grab a top30 and add it to the
original issue so we know what’s next on the hitlist?  Thanks

On Sep 20, 2018, at 7:34 PM, Jenny Buckley <notifications@github.com> wrote:

@smarterclayton <https://github.com/smarterclayton> About 5.5 million were
coming from the openapi endpoint handler, which creates just one full
openapi spec and serves it, and this change doesn't touch that code. I
think the openapi spec might just take that many objects to create.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/68576#issuecomment-423269048>,
or mute the thread

File: staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (mbohlool) [reply staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go]: can you reuse this please? these two copy can get out of sync. move them to somewhere common if necessary.

[the remark this answers is not in the document]
Comment (jennybuckley) [reply staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go]: What if I just put a TODO for this, and also make the PR to kube-openapi repo

**Comment by jennybuckley** [comment]
Since #70260 merged, this doesn't need to vendor in kube-openapi anymore. Should be good to merge now, uses the exported function from kube-openapi
cc @mbohlool

File: staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go [hunk 2]
```diff
@@ -445,3 +455,65 @@ func NewDefaultAPIGroupInfo(group string, scheme *runtime.Scheme, parameterCodec
...
+			if !pathsToIgnore.HasPrefix(path) {
+				kind, err := genericapi.GetResourceKind(groupVersion, storage, apiGroupInfo.Scheme)
+				if err != nil {
+					return gvkToSchema, err
+				}
+				sampleObjectPtr, err := apiGroupInfo.Scheme.New(kind)
+				if err != nil {
+					return gvkToSchema, err
+				}
+				sampleObject := reflect.Indirect(reflect.ValueOf(sampleObjectPtr)).Interface()
```
Comment (mbohlool) [inline staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go]: nit: you probably don't need this. it should work if you pass sampleObjectPtr.

**Comment by mbohlool** [comment]
only one nit comment. otherwise look good.

/lgtm

**Comment by jennybuckley** [comment]
Rebased

**Pull request merged**